server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +51,267 @@
   static final String DROPPED_COUNT = "droppedCount";
 
   @Override
-  public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntimeParams params, DataSegment segment)
+  public CoordinatorStats run(
+      final DruidCoordinator coordinator,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment
+  )
   {
+    final Object2IntMap<String> targetReplicants = new Object2IntOpenHashMap<>(getTieredReplicants());
+
+    final Object2IntMap<String> currentReplicants = new Object2IntOpenHashMap<>(
+        params.getSegmentReplicantLookup().getClusterTiers(segment.getIdentifier())
+    );
+
     final CoordinatorStats stats = new CoordinatorStats();
-    final Set<DataSegment> availableSegments = params.getAvailableSegments();
+    if (params.getAvailableSegments().contains(segment)) {
+      assign(targetReplicants, currentReplicants, params, segment, stats);
+    }
 
-    final Map<String, Integer> loadStatus = Maps.newHashMap();
+    drop(targetReplicants, currentReplicants, params, segment, stats);
 
-    int totalReplicantsInCluster = params.getSegmentReplicantLookup().getTotalReplicants(segment.getIdentifier());
-    for (Map.Entry<String, Integer> entry : getTieredReplicants().entrySet()) {
-      final String tier = entry.getKey();
-      final int expectedReplicantsInTier = entry.getValue();
-      final int totalReplicantsInTier = params.getSegmentReplicantLookup()
-                                              .getTotalReplicants(segment.getIdentifier(), tier);
-      final int loadedReplicantsInTier = params.getSegmentReplicantLookup()
-                                               .getLoadedReplicants(segment.getIdentifier(), tier);
+    return stats;
+  }
+
+  private static void assign(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment,
+      final CoordinatorStats stats
+  )
+  {
+    // if primary replica already exists
+    if (!currentReplicants.isEmpty()) {
+      assignReplicas(targetReplicants, currentReplicants, params, segment, stats, null);
+    } else {
+      final ServerHolder primaryHolderToLoad = assignPrimary(targetReplicants, params, segment);
+      if (primaryHolderToLoad == null) {
+        // cluster does not have any replicants and cannot identify primary holder
```

Comment:
Not really a problem _per se_, more like a "quick exit" because logically speaking, there is no need to further call `assignReplicasForTier` because none will be assigned.

Note that any alert/logging required will have been made during the `assignPrimary` call